File: android/app/src/main/java/com/example/app/domain/navcaddy/routing/RoutingResult.kt
package caddypro.domain.navcaddy.routing

import caddypro.domain.navcaddy.models.ParsedIntent
import caddypro.domain.navcaddy.models.RoutingTarget

/**
 * Result of routing orchestration.
 *
 * Based on intent classification, prerequisite validation, and intent type,
 * routing can result in:
 * - Navigate: Route user to target screen with state
 * - NoNavigation: Answer without screen change (query, help)
 * - PrerequisiteMissing: Required data missing, prompt user
 * - ConfirmationRequired: Mid-confidence intent needs confirmation
 *
 * Spec reference: navcaddy-engine.md R3
 */
sealed class RoutingResult {
    /**
     * Direct navigation to a target screen.
     * Used for high-confidence intents with all prerequisites met.
     *
     * @property target Routing destination with parameters
     * @property intent Original parsed intent for context
     */
    data class Navigate(
        val target: RoutingTarget,
        val intent: ParsedIntent
    ) : RoutingResult()

    /**
     * Answer without navigation.
     * Used for intents that don't require screen changes (pattern_query, help_request).
     *
     * @property intent Original parsed intent
     * @property response Answer to display to user
     */
    data class NoNavigation(
        val intent: ParsedIntent,
        val response: String
    ) : RoutingResult()

    /**
     * Prerequisites missing before routing can proceed.
     * User must provide required data first.
     *
     * @property intent Original parsed intent
     * @property missing List of missing prerequisites
     * @property message User-friendly explanation of what's needed
     */
    data class PrerequisiteMissing(
        val intent: ParsedIntent,
        val missing: List<Prerequisite>,
        val message: String
    ) : RoutingResult() {
        init {
            require(missing.isNotEmpty()) { "Must have at least one missing prerequisite" }
        }
    }

    /**
     * User confirmation required before routing.
     * Used for mid-confidence intents (0.50-0.74).
     *
     * @property intent Original parsed intent
     * @property message Confirmation question to present
     */
    data class ConfirmationRequired(
        val intent: ParsedIntent,
        val message: String
    ) : RoutingResult()
}

File: ios/App/Sources/Core/Domain/NavCaddy/Routing/RoutingResult.swift
import Foundation

/// Result of routing orchestration.
///
/// Spec R3: Routing orchestrator returns Navigate, NoNavigation, or PrerequisiteMissing.
enum RoutingResult {
    /// Navigate to target screen
    /// High-confidence intents route directly.
    case navigate(target: RoutingTarget, intent: ParsedIntent)

    /// Answer without navigation
    /// For informational intents like pattern_query, help_request.
    case noNavigation(intent: ParsedIntent, response: String)

    /// Prerequisite is missing
    /// Some intents require data before routing (e.g., recovery_check needs recovery data).
    case prerequisiteMissing(intent: ParsedIntent, missing: [Prerequisite], message: String)

    /// Confirmation required before routing
    /// Mid-confidence intents (0.50-0.74) need user confirmation.
    case confirmationRequired(intent: ParsedIntent, message: String)
}

extension RoutingResult {
    /// The parsed intent associated with this result
    var intent: ParsedIntent {
        switch self {
        case .navigate(_, let intent):
            return intent
        case .noNavigation(let intent, _):
            return intent
        case .prerequisiteMissing(let intent, _, _):
            return intent
        case .confirmationRequired(let intent, _):
            return intent
        }
    }

    /// Whether this result requires navigation
    var requiresNavigation: Bool {
        switch self {
        case .navigate:
            return true
        case .noNavigation, .prerequisiteMissing, .confirmationRequired:
            return false
        }
    }

    /// User-facing message if applicable
    var message: String? {
        switch self {
        case .navigate:
            return nil
        case .noNavigation(_, let response):
            return response
        case .prerequisiteMissing(_, _, let message):
            return message
        case .confirmationRequired(_, let message):
            return message
        }
    }
}
